### ISSUE TITLE
Allow `@MvcBinding` on classes

### ISSUE THREAD

**Issue by mvcbot** [eclipse-ee4j/krazo#26]
<a href="https://github.com/chkal"><img src="https://avatars2.githubusercontent.com/u/159245?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [chkal](https://github.com/chkal)**
_Sunday Aug 26, 2018 at 09:07 GMT_
_Originally opened as https://github.com/mvc-spec/ozark/issues/196_

----

As mentioned in mvc-spec/mvc-spec#161, it would be really nice to allow `@MvcBinding` on classes to simplify the creation of form classes used with `@BeanParam`. JAX-RS doesn't provide the required information. Maybe we can use some other trick to identify the `AnnotatedElement` for which the `ParamConverterProvider` is invoked so we can also check the class for annotations?

**Comment by erdlet** [comment]
Hey @chkal,

I just wanted to have a look on this and got a little bit confused, as it seems not to be possible even to test this without changing the MVC API itself, because `@MvcBinding` is only applicable to methods, parameters or fields. So is this really intended to be at the 1.0.0-m6 milestone, because in the Spec's issue it sounds like a feature which is targeted after a final 1.0 release? Or is there any new information not tracked here?

**Comment by chkal** [comment]
Well, I think the idea back then was to see if this can be implemented at all before putting it into the spec. I'm not fully sure if the JAX-RS parameter converter API supports this kind of requirement. And yes, if you want to experiment with this, you will need to update the spec API and build it yourself.

**Comment by erdlet** [comment]
Ok thanks for the information :+1: Think I'll have a look on this in the next days

**Comment by erdlet** [comment]
So I did some experiments on this issue and got a small (hacky) example with working `@MvcBinding` on class level [here](https://github.com/erdlet/krazo/tree/26-IDEA-DO-NOT-MERGE).

A simple integration test can be found [here](https://github.com/erdlet/krazo/blob/26-IDEA-DO-NOT-MERGE/testsuite/src/test/java/org/eclipse/krazo/test/BindingIT.java)

This is based on the Validation API instead of JAX-RS, because as mentioned by @chkal there seems no way to get the `@BeanParam` via JAX-RS API.

What's your opinion on this kind of solution (via Validation API)?

**Comment by erdlet** [comment]
So as a result, for conversion I didn't find anything which would allow to add `@MvcBinding` on a class with the current JAX-RS API.

To focus on the validation in the comment above was a little misunterstanding of the issue here from my side, sorry for that.

**Comment by chkal** [comment]
Hi @erdlet,

first of all: Thanks a lot for spending time to work on this. This is really a tricky issue and it would be great to support `@MvcBinding` on classes.

The code you wrote based on the validation API is interesting and seems to work in some cases. However, as you already mentioned, it seems like we cannot really get this to work correctly with the plain JAX-RS parameter converter API which would be a requirement to get this feature to work consistently across different use cases (binding AND validation errors). 

Unfortunately I doubt that we can get an improvement of the `ParamConverterProvider` API into JAX-RS. This API is very difficult to evolve without breaking backwards compatibility. But I'm open for ideas. :smile:

**Comment by hantsy** [comment]
if possible to make it work like this 

```
public Response save(@Valid @BeanParam @MvcBinding TaskForm form) 

```

The `@MvcBinding` can not be added on a POJO, but it can be added on a controller method para, but it does not work.

I have to set all fields in my bean to make it work.

```java
@RequestScoped
public class TaskForm implements Serializable {

    private static final long serialVersionUID = 1L;

    @NotBlank
    @FormParam("name")
    @MvcBinding
    private String name;

    @NotBlank
    @Size(min = 10, max = 2000)
    @FormParam("description")
    @MvcBinding
    private String description;
...
}
```

**Comment by chkal** [comment]
Yes, currently you have to annotate all the individual fields. And I agree that this is annoying. Unfortunately we are a limited to the annotations which are passed to our [ParamConverter](https://github.com/eclipse-ee4j/krazo/blob/1bde7ff1ddec2511c5cdb0534803ee2495655191/core/src/main/java/org/eclipse/krazo/binding/convert/MvcConverterProvider.java#L57) by the corresponding JAX-RS implementation.

However, improving this is still on my todo list. :-)

**Comment by matteoromanazzi** [comment]
> Yes, currently you have to annotate all the individual fields. And I agree that this is annoying. Unfortunately we are a limited to the annotations which are passed to our [ParamConverter](https://github.com/eclipse-ee4j/krazo/blob/1bde7ff1ddec2511c5cdb0534803ee2495655191/core/src/main/java/org/eclipse/krazo/binding/convert/MvcConverterProvider.java#L57) by the corresponding JAX-RS implementation.
> 
> However, improving this is still on my todo list. :-)

Hi, I'm trying with this hack, but still not working. On validation failed the server responds with 400 instead of invoking the post method. I'm using Tomee 8.0.10 with eclipse krazo 1.0.0 and the provided CXF for jax rs.

 this is the bean property

```
  @FormParam("text")
    @NotBlank
    @MvcBinding
    private String text;
```


this is the method

```
@POST
    public String post(@BeanParam @Valid DefineEventForm form){

        this.formState.setObj(form);
        return "redirect:/define-event";
    }
```

**Comment by chkal** [comment]
Hi @AlessioBandoni 

Unfortunately, Krazo isn't fully compatible with CXF. We even removed the CXF integration module some time ago. CXF shows some weird behavior, which IMO isn't even JAX-RS compatible. We created some issues in the CXF issue tracker, but they are still unresolved. Therefore, and because even OpenLiberty recently switched from CXF to RESTEasy, we decided to stop focusing on supporting CXF.

See #246 for details. Sorry about that.
